pkg/cmd/pulumi/crypto.go

```diff
@@ -54,11 +59,38 @@ func getStackDecrypter(s backend.Stack, ps *workspace.ProjectStack) (config.Decr
 	return dec, needsSave, nil
 }
 
-func getStackSecretsManager(s backend.Stack, ps *workspace.ProjectStack) (secrets.Manager, bool, error) {
+func getStackSecretsManagerFromState(ctx context.Context, s backend.Stack) (secrets.Manager, error) {
+	snap, err := s.Snapshot(ctx, stack.DefaultSecretsProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use the current snapshot secrets manager, if there is one, as the fallback secrets manager.
+	var defaultSecretsManager secrets.Manager
+	if snap != nil {
+		defaultSecretsManager = snap.SecretsManager
+	}
+	return defaultSecretsManager, nil
+}
+
+func getStackSecretsManager(
+	ctx context.Context, s backend.Stack, ps *workspace.ProjectStack,
+) (secrets.Manager, bool, error) {
+	// Try to get the secret manager from the stack's current state snapshot.
+	sm, err := getStackSecretsManagerFromState(ctx, s)
+	if err != nil {
+		return nil, false, err
+	}
+	if sm != nil {
+		// If we have a secrets manager from the snapshot, use it.
+		// Do not update the stack's configuration.
+		// TODO: Consider warning if the snapshot secrets manager is different from the
```

Comment:
My thinking was that config takes priority, if there is no config file then use what's in state. If there is a config file but it doesn't match whats in state than give a warning because it probably means config was changed but "stack change-secrets-provider" wasn't used to do it correctly.

Reply:
So if I want to change secret provider, will I be able to do it using a simple edit of the config file?

I think maybe just actually block it, until `change-secrets-provider` was called explicitly, i.e. not a warning, but an error.

If someone accidentally does some `git` games, restoring old config files into the working directory and now it does not match the current secret provider, I think it would be wise to block any accidental changes.

No?